### PR TITLE
Add sigmoid to lightgbm parameter list and change default to 1.

### LINKED ISSUE #654: Default value of sigmoid parameter for LightGBM
[The default value of the `sigmoid` parameter for LightGBM is 1](https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst#sigmoid), however in the ML.Net wrapper it is hardwired to 0.5:
https://github.com/dotnet/machinelearning/blob/f9d3973a056ad26bc6cc15c2d7a09f8ae47e30da/src/Microsoft.ML.LightGBM/LightGbmArguments.cs#L404

Why is this the case? Shouldn't this parameter either be 1 or exposed on the ML.Net API instead?

### PR DESCRIPTION
**Pull request by codemzs**
fixes #654

### CODE REVIEW

**Comment by codemzs** [comment]
@guolinke Please take a look at this PR and let me know if it looks good to you.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 1]
```diff
@@ -401,7 +406,7 @@ public enum EvalMetricType
             }
             if (!string.IsNullOrEmpty(metric))
                 res["metric"] = metric;
-            res["sigmoid"] = "0.5";
+            res["sigmoid"] = Sigmoid.ToString();
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >Sigmoid [](start = 29, length = 7)

must be greater than zero according to [https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst](https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 2]
```diff
@@ -321,6 +321,11 @@ public enum EvalMetricType
         [TGUI(Label = "Ranking Label Gain")]
         public string CustomGains = "0,3,7,15,31,63,127,255,511,1023,2047,4095";
 
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Parameter for the sigmoid function. Used only in binary and multiclassova " +
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >multiclassova [](start = 115, length = 13)

This is a nice description for direct instantiation of lightgbm, in our case it need to be translated to multiclass trainer with useSoftMax = true

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 3]
```diff
@@ -321,6 +321,11 @@ public enum EvalMetricType
...
+            "classification and in lambdarank applications", ShortName = "sigmoid")]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >lambdarank [](start = 35, length = 10)

just ranking.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 1 above)
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >.ToString(); [](start = 36, length = 12)

Please, no ToString for this parameter, and any other in this method (except EvalMetric), also please remove, ToString from UpdateParameters in LightGbmArguments.cs  in field.GetValue(Args).ToString()

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 3 above)
Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: lambda rank is the ranking objective function.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 2 above)
Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Can you clarify why?

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 3 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Lambda rank and multiclassova is internal Lightgbm terminology. Consider user who never heard about lightgbm, but want to give it a try. For this user we don't expose anything related to lambda ranking or multiclassova, we have our own terminology of ranking trainer, or multiclassification, and we should use it.

---
In reply to: [208723780](https://github.com/dotnet/machinelearning/pull/667#discussion_r208723780) [](ancestors = 208723780)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: this is only place where we expose this "multiclassova" for user, and it's unclear what it's means. if you want user to read documentation on lightgbm github, you can just put link to internal lightgbm documentation here, but I think it would be somewhat rude.

---
In reply to: [208724218](https://github.com/dotnet/machinelearning/pull/667#discussion_r208724218) [](ancestors = 208724218)

Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: It means multi-class ova. Seems pretty clear to me.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 3 above)
Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: LamdaRank is a famous ranking algorithm, it is not just used in LightGBM but also in fasttree. Read more here: https://www.microsoft.com/en-us/research/publication/from-ranknet-to-lambdarank-to-lambdamart-an-overview/

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: I have no doubt what lambdarank is famous algorithm, it's more a question of how often we use it in any user facing api?
We do have multiclass classification, binary classification, ranking, we have LightGbmRanker class, but how often user see word lambdarank? 
Look from user perspective, are they same, or there is difference.

All I'm trying to say, I don't understand what's the point of copying documentation from lightgbm site which is confusing in ML.Net (this is two different projects) if we can just rephase it in more ML.net appropriate terminology.

---
In reply to: [208727880](https://github.com/dotnet/machinelearning/pull/667#discussion_r208727880) [](ancestors = 208727880)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Ok, I'm doing something like
pipeline.Add(new LightGbmClassifier(...));
Is it multiclassova or not? 
Should I set sigmoid to something other than default value or it's useless?

Or should I do pipeline.Add(new Ova(LightGbmBinaryClassifier)) instead?
Is this multiclassova?

---
In reply to: [208727571](https://github.com/dotnet/machinelearning/pull/667#discussion_r208727571) [](ancestors = 208727571)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 3 above)
Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Your concern as I understand from your second comment on this thread has to do with "Lambda rank and multiclassova is internal Lightgbm terminology" and as I have pointed out previously this statement is not accurate. 

Now lets talk about your other concerns that you have brought up in your most recent comment, I think we should change the documentation where there is a need to and we should also try to keep the parameter list and documentation as close to LightGBM's official documentation so that our users that have been using command line and python implementation of lightGBM are not confused and its easy for them. Given there are [so many](https://en.wikipedia.org/wiki/Learning_to_rank) ranking algorithms out there it is better to be specific about which one we are exposing than having user ask questions or figure themselves out by reading code or however.

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: pipeline.Add(new LightGbmRanker(...));
Can I use sigmoid in arguments, or it's useless?
Is it lambdarank or not? And more specifically, how user find answer on this question?

---
In reply to: [208734818](https://github.com/dotnet/machinelearning/pull/667#discussion_r208734818) [](ancestors = 208734818)

**Comment by Ivanidzo4ka** [comment]
>             res["seed"] = host.Rand.Next().ToString();

These 3 calls of ToString completely unnecessary.

---
Refers to: src/Microsoft.ML.LightGBM/LightGbmArguments.cs:379 in f2e5485. [](commit_id = f2e54852d2d2a77180c3eee75317b93c683ef6c9, deletion_comment = False)

**Comment by justinormont** [comment]
Have benchmarks been run on this change? I think changing the sigmoid hyperparameter should have an effect. Without testing I don't know that 0.5 is better or worse than 1.0. @guolinke, how much benchmarking was done on your side?

In general, we are well setup to benchmark these changes and we don't know the extent which the source ML algos were benchmarked, so I don't think we should automatically echo the default hyperparameters set upstream. Automated daily regression tests are a stronger answer to changes like this, though I think kicking off a manual test is appropriate in the meantime.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 4]
```diff
@@ -321,6 +321,11 @@ public enum EvalMetricType
...
+        [TGUI(Label = "Sigmoid")]
+        public double Sigmoid = 1;
```
Comment (justinormont) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: If the hyperparameter has a notice effect, and the user can generally get gains by changing it, we should add a sweep range to the hyperparameter. E.g.: https://github.com/dotnet/machinelearning/blob/f2e54852d2d2a77180c3eee75317b93c683ef6c9/src/Microsoft.ML.LightGBM/LightGbmArguments.cs#L135-137

Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Do we have a sweep range for sigmoid that you are aware of?

Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Do you mean between 0 and 1? (not including zero). @guolinke your thoughts?

Comment (justinormont) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: That's a good question. 

As we had a fixed value before, I haven't had a chance to explore what a good range looks like. Values >0 is acceptable, but I don't know "good" ranges; for instance is 10 a useful value? 500? Or should we focus (0..1]?

Comment (guolinke) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Not sure about the effect of sigmoid. smaller sigmoid somehow like the using smaller learning rate. So I think using learning_rate searching may is enough.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 5]
```diff
@@ -401,13 +407,13 @@ public enum EvalMetricType
             }
             if (!string.IsNullOrEmpty(metric))
                 res["metric"] = metric;
-            res["sigmoid"] = "0.5";
```
Comment (justinormont) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Have benchmarks been run on this change? I think changing the sigmoid hyperparameter should have an effect. Without testing I don't know that 0.5 is better or worse than 1.0. @guolinke, how much benchmarking was done on your side?

In general, we are well setup to benchmark these changes and we don't know the extent which the source ML algos were benchmarked, so I don't think we should automatically echo the default hyperparameters set upstream. Automated daily regression tests are a stronger answer to changes like this, though I think kicking off a manual test is appropriate in the meantime.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 2 above)
Comment (TomFinley) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >multiclassova [](start = 115, length = 13)

Two points:

First, consider how FastTree is arranged. If an argument is only relevant to some trainer, then the way we accomplish that is by having that argument appear only in those specific argument classes that it is actually relevant. Is that possible to do here? This arrangement of "don't use this argument unless you're using this trainer" is really awkward, and not well arranged.

Even failing that, this word "multiclassova" is not one that I see is exposed to the user anywhere else, and we ought to therefore use a better description. `LightGbmMulticlassTrainer` perhaps. If you do that, please do `nameof` or something.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 3 above)
Comment (TomFinley) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Reactivating. For intellisense, it would be better if the argument isn't relevant to that trainer that it simply not appear *at all*.

---
In reply to: [208740753](https://github.com/dotnet/machinelearning/pull/667#discussion_r208740753) [](ancestors = 208740753)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 5 above)
Comment (codemzs) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: I have reverted the default. If someone wants to use sigmoid=1 they can set it now.

Comment (rogancarr) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: This *will* have a change, but if it's the default value in the `LightGBM` codebase, then I would suggest we keep it as 0.5. This is adding regularization to the gradient, whereas a value of `1` is basically adding no regularization.

@guolinke Was the value of `0.5` used to cancel out the factor of `2` used in the LambaRank-style loss? This value of `0.5` seems like it was put in to make sure the classifier learned the exact Logit. Does that sound right?
